### PR TITLE
Update UPSERT comment now that native upserts are the default

### PR DESCRIPTION
**Pull request by DMRobertson**
Follow-up to https://github.com/matrix-org/synapse/pull/13760#issuecomment-1251129040. I have tried to write down my understanding and better explain what the heck is going on. It may be too verbose for your taste.

### CODE REVIEW

**Review by babolivier — APPROVED** [review]
I'm always in favour of verbose comments

File: synapse/storage/database.py [hunk 1]
```diff
@@ -1141,17 +1141,48 @@ async def simple_upsert(
...
-            1. there is a UNIQUE INDEX on the key columns. In this case a conflict
-            will cause an IntegrityError in which case this function will retry
-            the update.
-            2. we somehow know that we are the only thread which will be updating
-            this table.
-        As an additional note, this parameter only matters for old SQLite versions
-        because we will use native upserts otherwise.
+        [*]: This class is aware that some tables have unique indices added as
+             background updates. It checks at runtime to see if those updates are
+             pending: if not, they are deemed safe for use with native upserts.
```
Comment (clokep) [inline synapse/storage/database.py:1185]: Is it worth pointing to the mapping of these? Otherwise it is left feeling like magic.

Comment (DMRobertson) [reply synapse/storage/database.py:1185]: (The whole thing does feel somewhat magical to me, tbh; hence the comment.)

I'll try to get this done today.

Comment (clokep) [inline synapse/storage/database.py:1185]: Maybe something like:

```suggestion
        [*]: At runtime this class checks the tables/background indices in
        	UNIQUE_INDEX_BACKGROUND_UPDATES to see if it is safe to
        	do native upsert operations on those tables. It continually checks
        	until all associated background updates are finished.

        	Note that if a table doesn't appear in that list it is assumed to be
        	safe for native upserts.
```

Comment (DMRobertson) [reply synapse/storage/database.py]: I've had a go in c46f103 based on some of your words. WDYT?

**Review by clokep — APPROVED** [review]
Thanks!

**Comment by DMRobertson** [comment]
(Manually merging because "screw you, test flakes")

**Pull request merged**